### PR TITLE
health-bar performance + visual ( by Patela )

### PR DESCRIPTION
**Pull request by P4tela**
Problems that were fixed:
     Visual           -> Some times the bars wouldn't be visible for the all match(you had to relog to fix it)
     Performance -> The data and position from each bar were being loaded/calculated each 0,3 seconds ( per account )
     Bar Data       -> The background color of each bar wasn't coherent ( if shields were present )

Patela

### CODE REVIEW

**Comment by JulienCoutault** [comment]
Why do you remove the possibility to change the shotcut ?

File: src/app/mods/health-bar/health-bar.ts [hunk 1]
```diff
@@ -1,153 +1,148 @@
...
-            }
+    startMod(): void {}
 
-            .lifeBar {
-                transition-duration: 300ms;
-                height: 100%;
-                width: 0%;
-                background-color: #333;
-            }
+    private bars: { [fighterId: number]: Bar; } = { };
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:10]: Hoist class members to the top of the class

File: src/app/mods/health-bar/health-bar.ts [hunk 2]
```diff
@@ -1,153 +1,148 @@
...
 
+        .lifeBarContainer {
+            box-sizing: border-box;
+            background-color: #222;
+            height: 6px;
+            width: 80px;
+            position: absolute;
+            border-radius: 3px;
+            overflow: hidden;
+            transition-duration: 500ms;
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:46]: Duration but no transition-property ?

File: src/app/mods/health-bar/health-bar.ts [hunk 3]
```diff
@@ -1,153 +1,148 @@
...
+            margin-top: 10px;
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:47]: Item being absolutely positioned, use "top" and not "margin-top"

File: src/app/mods/health-bar/health-bar.ts [hunk 4]
```diff
@@ -1,153 +1,148 @@
...
+        }
 
-            this.shortcutsHelper.bind(this.params.health_bar_shortcut, () => {
-                this.barContainer.toggle();
-            });
+        .lifeBar {
+            transition-duration: 300ms;
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:51]: No property along with duration

File: src/app/mods/health-bar/health-bar.ts [hunk 5]
```diff
@@ -1,153 +1,148 @@
...
         }
-    }
 
+        .lifePointsText {
+            font-size: 12px;
+            position: absolute;
+            width: 80px;
+            color: white;
+            text-shadow: 0 0 5px rgba(0, 0, 0, 0.9);
+            transition-duration: 500ms;
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:63]: No property along with duration

File: src/app/mods/health-bar/health-bar.ts [hunk 6]
```diff
@@ -1,153 +1,148 @@
...
+            margin-top: 16px;
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:64]: Use "top" and "left", not margin

File: src/app/mods/health-bar/health-bar.ts [hunk 7]
```diff
@@ -1,153 +1,148 @@
...
-    private setFightStart(): void {
-        this.on(this.wGame.dofus.connectionManager, 'GameFightStartingMessage', () => {
-            try {
-                this.fightJustStarted = true;
-            } catch (ex) {
-                Logger.error(ex);
-            }
-        });
+    private createHealthBars(){
+        if (this.wGame.document.getElementById('lifeBars') != null) return;
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:86]:
```js
if (this.wGame.document.getElementById('lifeBars')) return;
```

File: src/app/mods/health-bar/health-bar.ts [hunk 8]
```diff
@@ -1,153 +1,148 @@
...
-                if (this.fightJustStarted) {
-                    this.fightJustStarted = false;
-                    this.barContainer.fightStarted();
-                }
-            } catch (ex) {
-                Logger.error(ex);
-            }
-        });
+    private showHealthBars(){
+        if (this.container.style.visibility != "visible"){
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:97]: This implies the style HAS to be set with JS. Instead, use `window.getComputedStyle(this.container).visibility`.

File: src/app/mods/health-bar/health-bar.ts [hunk 9]
```diff
@@ -1,153 +1,148 @@
...
-        super.reset();
-        if (this.params.health_bar) {
-            if (this.shortcutsHelper) this.shortcutsHelper.unBindAll();
-            if (this.barContainer) this.barContainer.destroy();
-            const healthbarCss = this.wGame.document.getElementById('healthbarCss');
-            if (healthbarCss && healthbarCss.parentElement) healthbarCss.parentElement.removeChild(healthbarCss);
+    private destroyHealthBar(fighterId: any){
+        if (this.bars[fighterId]) {
+            this.bars[fighterId].destroy();
+            delete this.bars[fighterId];
```
Comment (Maryannah) [inline src/app/mods/health-bar/health-bar.ts:137]: Delete removes the content of a memory reference, but not the memory reference itself. Instead, use `Array.prototype.splice`

**Review by Maryannah — CHANGES_REQUESTED** [review]

File: src/app/mods/health-bar/health-bar.ts (hunk 6 above)
Comment (P4tela) [reply src/app/mods/health-bar/health-bar.ts:64]: Doesn't make sense... Top behavior usually depends on the position and in this case, using "top" I won't get the result I'm looking for

File: src/app/mods/health-bar/health-bar.ts (hunk 9 above)
Comment (P4tela) [reply src/app/mods/health-bar/health-bar.ts:137]: I don't have a lot of experience in this language, so do you mind writing the code i should use to replace "delete this.bars[fighterId];"? 
I wasn't able to fully understand your point...

File: src/app/mods/health-bar/health-bar.ts (hunk 5 above)
Comment (P4tela) [reply src/app/mods/health-bar/health-bar.ts:63]: Object changes position so the transition-duration is used(i changed the value for 300 tho)

File: src/app/mods/health-bar/health-bar.ts (hunk 2 above)
Comment (P4tela) [reply src/app/mods/health-bar/health-bar.ts:46]: Object changes position so the transition-duration is used(i changed the value for 300 tho)

**Review by P4tela — COMMENTED** [review]
let me know if you prefer takling on discord instead

File: src/app/mods/health-bar/health-bar.ts (hunk 5 above)
Comment (Maryannah) [reply src/app/mods/health-bar/health-bar.ts:63]: Leaving the transition property empty will result in inconsistent behaviors, either no animation, or animations on properties you do not want to animate

File: src/app/mods/health-bar/health-bar.ts (hunk 6 above)
Comment (Maryannah) [reply src/app/mods/health-bar/health-bar.ts:64]: Position is absolute, meaning the effect is the same as the margin. As a good practice, use the position properties to position the elements, and use the boxing properties to adjust spacing between elements.

Comment (P4tela) [reply src/app/mods/health-bar/health-bar.ts:64]: I tried to remove to leave the "top" and "left" ( aka removing the "margin-*") and the result is not the same. that's why i said that....Should i be doing anything else besides removing the "margin-*"?

Comment (Clover-Lindo) [reply src/app/mods/health-bar/health-bar.ts:64]: In that specific case, it won't be possible to switch to "top" and "left", as the JS code already uses this property to position the element above the canvas (if I'm not mistaken).
To offset the final position, we'll have to either use margins, translates, or to edit the position computed by the JS part.
I haven't yet run the change on Lindo to see the desired result though, there may be a better solution.

File: src/app/mods/health-bar/health-bar.ts (hunk 8 above)
Comment (P4tela) [reply src/app/mods/health-bar/health-bar.ts:97]: this property is always visible. not what I want

File: src/app/mods/health-bar/health-bar.ts (hunk 6 above)
Comment (P4tela) [reply src/app/mods/health-bar/health-bar.ts:64]: Does this mean I can leave it the way it is?

**Comment by JulienCoutault** [comment]
> > Why do you remove the possibility to change the shortcut ?
> 
> Having a checkbox and a shortcut seemed useless. Only one of them should be enough...

The health-bar mode has 3 states : off/bar only/bar + pv
The option can just hide/show the mod, the shortcut permit to switch the two last state.
And people like shortcut to fastly enable/disable feature without checking settings

**Comment by P4tela** [comment]
> 
> 
> > > Why do you remove the possibility to change the shortcut ?
> > 
> > 
> > Having a checkbox and a shortcut seemed useless. Only one of them should be enough...
> 
> The health-bar mode has 3 states : off/bar only/bar + pv
> The option can just hide/show the mod, the shortcut permit to switch the two last state.
> And people like shortcut to fastly enable/disable feature without checking settings

Ye I talked with some users and that didn't seem to be a problem at all(turn off the option by using the check-box). I thought of leaving the shortcut but seemed useless to have both options. 
Plus, having an option where you can only see the bar isn't really usefull (imo). 
If everyone thinks it's usefull and it's used more than once per day then I'll add it. If not, then we'll be able to save a key slot ( for the shortcuts) and spare the program with useless features(code/logic).

File: src/app/mods/health-bar/health-bar.ts (hunk 9 above)
Comment (Clover-Lindo) [reply src/app/mods/health-bar/health-bar.ts:129]: @Maryannah Isn't splice meant for Arrays? `this.bars` is an object here

File: src/app/mods/health-bar/health-bar.ts (hunk 6 above)
Comment (Maryannah) [reply src/app/mods/health-bar/health-bar.ts:55]: Then a good practice would be to group the style altogether, or at least leave comments to explain the use case here !

File: src/app/mods/health-bar/health-bar.ts (hunk 8 above)
Comment (Maryannah) [reply src/app/mods/health-bar/health-bar.ts:89]: The code I provided gives the actual value of the property. The code you use gives the value provided in the "style" attribute of the tag (which can be overridden with a "!important" annotation).

File: src/app/mods/health-bar/health-bar.ts (hunk 9 above)
Comment (Maryannah) [reply src/app/mods/health-bar/health-bar.ts:129]: If so my bad then, ignore this comment !

**Comment by JulienCoutault** [comment]
You're maybe right for the option with only the bar, we should discuss about it.
For the shortcut, lot of people asked for more shortcut to enable/disable feature, so we must keep it

**Comment by Clover-Lindo** [comment]
image

**Comment by Clover-Lindo** [comment]
So there is around 6% of users, on a sample of 100+ users, who are making use of this shortcut.
It's a minority, but it may justify to keep the shortcut with the simplified displayed.

What do you think?

**Comment by P4tela** [comment]
> 
> 
> image

I assume the number 5 is something that is actually used cus of the visual bug on the text. If the chars were displayed in a certain position, the numbers will position themselfs behind the health-bars ( maybe 20% of the text(bottom horizontal part)). I also fixed that which was annoying me.
But again, it's not my call. I think It's useless to have it since only +/-10 people out of +/-160 people use it.
If it's something that you want me to add then please let me know.

**Comment by Maryannah** [comment]
Commenting on this matter : you usually do not remove a feature that has no issues. Even more when it's a feature that is currently in use.

This is only a personal opinion, but as a rule of thumb, if there is no problem, don't touch it !

**Comment by P4tela** [comment]
I would agree with you if this was a normal situation. But this one isn't really normal... I saw a feature that in my 2 years of non-stop gameplay i never, ever, used. Me and several others from my guild. So having useless stuff in your code is never usefull. That's why i asked if anyone was actually using it since i never saw people that were. But I also said that if I'm wrong or if it's something that the admins want to keep then I'll re-add it without problems. But re-adding it cus "it was already there and it worked" but nobody uses it then it's nonsense...

**Comment by Maryannah** [comment]
You are not everyone, you are a sample of everyone. 6% of another sample use it. So you're not removing "something useless", you're removing something that some people do not use. This is a huge difference.

Nobody asked for it to be removed, and this is not an improvement (neither from a user experience, nor a code quality point of view).

Anyway, you seem to be a bit stubborn, and I'm not in the decision making process, I was just giving my opinion from a professional point of view, so I'll stop here about this matter and let you all finish it

**Comment by Clover-Lindo** [comment]
Thanks you both for sharing your opinion. I'll wait for a few more, maybe we can get some other interresting point of views.

**Comment by P4tela** [comment]
I may have expressed myself poorly because i didn't want to pass the idea that if i don't use it then nobody uses it...
I just wanted to say that if the feature is needed then I'll add it.

**Comment by Clover-Lindo** [comment]
Interresting feedback from a user that did not know about this feature, and thought it was a bug

image
image

**Comment by Clover-Lindo** [comment]
To sum up the feedback I got on Discord:

The keyboard shortcut is often used as a quick solution to refresh the bars when some bugs occur:
- If a player is invisible when joining a fight already in progress, the bar will not be created once the player becomes visible again
- Sometimes, the HP number or the whole health bars are not displayed, and not after requesting it deliberately

This last point seems to be an involuntary press of the keyboard shortcut, but a player reported the chances of this to be extremely low for him, meaning it may be due to something else.

Also, the main point behind having only the bars displayed without the HP number was to simplify the UI when to much details are already present on the screen.

---

Given these feedbacks, here is what I think:
**Removing the HP number** seems optional to me, it's a cool plus in some cases, but it is also often perceived as a bug. So this option **should be removed**.
The shortcut is used for a different purpose than its initial purpose, and until we can guarantee these bugs are fixed, **we should keep this shortcut** to avoid frustration. And this may be used as a work around to simplify the UI when there are too much details on the screen.
Finally, having the shortcut being set should not be considered as a defaut behavior given the amount of players not using it, **the selected key should be blank by default**.

What do you think?

**Comment by P4tela** [comment]
The bugs did happen several times (at least for me) but those are fixed now.
About the shortcut... in my opinion if we're going to leave the option to hide the health-bar + the text just because of a low amount of people then it's logical to leave the option to hide just the text for another low amount of people...

Thanks agains for collecting all the information Clover! :)

**Comment by Clover-Lindo** [comment]
> The bugs did happen several times (at least for me) but those are fixed now.

That's great 👍 

> if we're going to leave the option to hide the health-bar + the text just because of a low amount of people then it's logical to leave the option to hide just the text for another low amount of people

Aside the "low amount of people" argument, is there any other similarity that makes you think two different things should necessarily share the same decision?

**Comment by P4tela** [comment]
Let's say we have option A and option B.
Both options were "removed" because of the same reason and both options are specific features within a feature that are exceptions. If we're going to satisfy the users by adding option A then why not add option B ? 
If we listen to those 9 then why not listen to the other 7?
imagem

And yes, this might be weird comming from the guy that removed it in the first place. But if the decision was mine alone then i would remove both. Just adding one doesn't make sense...

**Comment by Clover-Lindo** [comment]
I've come to this by taking users feedback into account.

On the one hand, you have a feature that is percieved as a bug by maybe more than the ones who are actually using this feature: The ability to remove the HP number.
On the other hand, we have a feature that provides a workaround for bugs around the health bar feature, that also provides a workaround for the other feature in case we remove it (which means we're combining 4 and 5 from the survey), and that also still fulfill its original purpose: Being able to quickly remove the health bar display when they become too much nois in some circumstances.

I'll be quoting Maryannah on this one:

> you usually do not remove a feature that has no issues

- Having the bars alone being displayed is percieved as a bug
- The keyboard shortcut fulfills its duty and can be used as a bug workaround

So for now, if there is no argument other than "Both features are used by the same amount of people so we should treat them equally" then it's going to be hard for me to change my mind. I really believe we should consider each feature individually and weight the pros and cons of both.

That being said, I understand it may bring an additional complexity to remove only one of both, as it will require taking care of backward compatibility with existing settings.

**Comment by P4tela** [comment]
Okay, I didn't thought of that. That does make sense and I can get behind that. 
I already removed the one you wanted me to remove (you can check the latest commit).

**Comment by Clover-Lindo** [comment]
Whenever you are confident enough about the stability we can merge this into the master branch 👍

**Comment by P4tela** [comment]
I didn't removed the event cus i didn't seem necessary, to remove and add every single time the user uses the shortcut. So i left the event always running and he doesn't run the method inside of the event if the shortcut was previously toggled. 
If you agree then you can merge it.

**Comment by Clover-Lindo** [comment]
I've been running some tests and noticed a few things:
- When the mod is reseted, the container is not removed from the document, making newest instances of the mod believe the container has already been initialized (but not affected to this.container)
- When starting a fight, the standardHealthBars method is called but always ignored because:
  - The #lifeBars element is already present on the document
  - Either `this.container.style.visibility` is `'visible'` or `this.visibleBars` is `false`
- Even if the call is always ignored, the `showHealthBars` method should never be called more than once. This would intialize redundant event listeners that would be triggered twice, three times and so on.
- `this.container.style.visibility` here is used as a variable, because the inline style is either equal to `'visible'` or not set, which falls back to `'visible'` since this is the default value. The style has no visual incidence.

I know this PR has now been open for a long time and I don't want to personally go through multiple reviews & tests to check the fixes for each point so I will be pushing the fixes directly and merging this into the original repo. This will prevent us going through multiple back and forth for the next days.

**Pull request merged**